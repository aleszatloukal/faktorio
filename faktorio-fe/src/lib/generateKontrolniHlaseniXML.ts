import { Invoice } from '@/components/IssuedInvoiceTable'
import { ReceivedInvoice } from '@/components/ReceivedInvoiceTable'
import { formatCzechDate, formatXmlNumber } from './utils'

export interface SubmitterData {
  dic: string
  typ_ds: string
  jmeno: string
  prijmeni: string
  ulice: string
  psc: string
  stat: string
  email: string
  // Add other fields if needed from the query
}

interface GenerateXmlParams {
  issuedInvoices: Invoice[]
  receivedInvoices: ReceivedInvoice[]
  submitterData: SubmitterData
  year: number
  quarter?: number
  month?: number
}

export function generateKontrolniHlaseniXML({
  issuedInvoices,
  receivedInvoices,
  submitterData,
  year,
  quarter,
  month
}: GenerateXmlParams): string {
  const VAT_THRESHOLD = 10000 // CZK threshold for B2/B3 split
  const todayCzech = formatCzechDate(new Date())

  // Process Issued Invoices (VetaA4)
  let vetaA4Xml = ''
  let issuedInvoiceSubtotalSum = 0
  issuedInvoices.forEach((inv, index) => {
    const clientVatId = inv.client_vat_no || 'MISSING_DIC_ODB'
    const taxableDate = formatCzechDate(inv.taxable_fulfillment_due)
    const subtotal = inv.subtotal ?? 0
    const vatAmount = (inv.total ?? 0) - subtotal
    issuedInvoiceSubtotalSum += subtotal

    if (
      !inv.number ||
      !taxableDate ||
      !clientVatId ||
      clientVatId === 'MISSING_DIC_ODB'
    ) {
      console.warn(
        'Skipping issued invoice due to missing data (number, taxableDate, clientVatId):',
        inv
      )
      return
    }

    vetaA4Xml += `
    <VetaA4
      c_radku="${index + 1}"
      dic_odb="${clientVatId.replace('CZ', '')}"
      c_evid_dd="${inv.number}"
      dppd="${taxableDate}"
      zakl_dane1="${formatXmlNumber(subtotal)}"
      dan1="${formatXmlNumber(vatAmount)}"
      kod_rezim_pl="0"
      zdph_44="N"
    />`
  })

  // Process Received Invoices (VetaB2 and VetaB3)
  let vetaB2Xml = ''
  let b3TotalSubtotal = 0
  let b3TotalVat = 0
  let receivedInvoiceSubtotalSum = 0
  let b2Index = 0

  receivedInvoices.forEach((inv) => {
    const supplierVatId = inv.supplier_vat_no || 'MISSING_DIC_DOD'
    const taxableDate = formatCzechDate(inv.issue_date)
    const subtotal = inv.total_without_vat ?? 0
    const totalWithVat = inv.total_with_vat ?? 0
    const vatAmount = totalWithVat - subtotal
    receivedInvoiceSubtotalSum += subtotal

    if (
      !inv.invoice_number ||
      !taxableDate ||
      !supplierVatId ||
      supplierVatId === 'MISSING_DIC_DOD'
    ) {
      console.warn(
        'Skipping received invoice due to missing data (invoice_number, taxableDate, supplierVatId):',
        inv
      )
      return
    }

    if (totalWithVat > VAT_THRESHOLD && inv.currency === 'CZK') {
      b2Index++
      vetaB2Xml += `
    <VetaB2
      c_radku="${b2Index}"
      dic_dod="${supplierVatId.replace('CZ', '')}"
      c_evid_dd="${inv.invoice_number}"
      dppd="${taxableDate}"
      zakl_dane1="${formatXmlNumber(subtotal)}"
      dan1="${formatXmlNumber(vatAmount)}"
      zdph_44="N" 
      pomer="N"
    />`
    } else {
      b3TotalSubtotal += subtotal
      b3TotalVat += vatAmount
    }
  })

  const vetaB3Xml = `
    <VetaB3
      zakl_dane1="${formatXmlNumber(b3TotalSubtotal)}"
      dan1="${formatXmlNumber(b3TotalVat)}"
    />`

  // Construct Final XML
  let periodAttribute = ''
  if (quarter !== undefined) {
    periodAttribute = `ctvrt="${quarter}"`
  } else if (month !== undefined) {
    const formattedMonth = month.toString()
    periodAttribute = `mesic="${formattedMonth}"`
  } else {
    throw new Error(
      'Either month or quarter must be provided for XML generation.'
    )
  }

  const xmlString = `<?xml version="1.0" encoding="UTF-8"?>
<Pisemnost nazevSW="EPO MF ÄŒR" verzeSW="41.16.3">
<DPHKH1 verzePis="03.01">
  <VetaD k_uladis="DPH" dokument="KH1"
    rok="${year}" ${periodAttribute}
    d_poddp="${todayCzech}"
    khdph_forma="B"
  />
  <VetaP
    dic="${submitterData.dic.replace('CZ', '')}" typ_ds="${submitterData.typ_ds}" jmeno="${submitterData.jmeno}" prijmeni="${submitterData.prijmeni}" ulice="${submitterData.ulice}" psc="${submitterData.psc}" stat="${submitterData.stat}" email="${submitterData.email}" sest_jmeno="${submitterData.jmeno}" sest_prijmeni="${submitterData.prijmeni}"
  />
  ${vetaA4Xml}
  ${vetaB2Xml}
  ${vetaB3Xml}
  <VetaC
    obrat23="${formatXmlNumber(issuedInvoiceSubtotalSum)}"
    pln23="${formatXmlNumber(receivedInvoiceSubtotalSum)}"
  />
</DPHKH1>
</Pisemnost>`

  return xmlString
}
