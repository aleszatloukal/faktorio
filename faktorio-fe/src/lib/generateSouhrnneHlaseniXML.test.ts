import { describe, it, expect, vi } from 'vitest'
import { SubmitterData } from './generateKontrolniHlaseniXML'
import { generateSouhrnneHlaseniXML } from './generateSouhrnneHlaseniXML'

// Define a type for the test invoice structure matching tRPC result expectation
type TestInvoice = {
  id: string
  invoice_number?: string | null
  customer_name?: string | null
  client_vat_no?: string | null
  issue_date?: string | null
  due_date?: string | null
}

describe('generateSouhrnneHlaseniXML', () => {
  it('should generate correct XML using correct field names', () => {
    const submitterData: SubmitterData = {
      dic: 'CZ8807204153',
      typ_ds: 'F',
      prijmeni: 'Špác',
      jmeno: 'Jiří',
      naz_obce: 'SOKOLNICE',
      ulice: 'Třešňová',
      c_pop: '823', // Added house number
      psc: '66452',
      stat: 'ČESKÁ REPUBLIKA',
      email: 'test@example.com'
    }

    const issuedInvoices = [
      {
        id: '1',
        invoice_number: '2025-001',
        customer_name: 'Test CZ Client',
        client_vat_no: 'CZ12345678',
        issue_date: '2025-03-01',
        native_total: 12100, // CZK invoice, should be ignored
        currency: 'CZK'
      },
      {
        id: '2',
        invoice_number: '2025-002',
        customer_name: 'First DE Client GmbH',
        client_vat_no: 'DE360131145',
        issue_date: '2025-02-28',
        native_total: 76388.8125, // Use native_total
        currency: 'EUR' // EUR invoice, should be included
      },
      {
        id: '3',
        invoice_number: '2025-003',
        customer_name: 'Another DE Client GmbH',
        client_vat_no: 'DE987654321',
        issue_date: '2025-03-10',
        native_total: 25000.5, // Use native_total
        currency: 'EUR'
      },
      {
        id: '4',
        invoice_number: '2025-004',
        customer_name: 'French Client SARL',
        client_vat_no: 'FR12345678901',
        issue_date: '2025-03-15',
        native_total: 50000.0, // Use native_total
        currency: 'EUR'
      }
    ]

    const year = 2025
    const quarter = 1

    const result = generateSouhrnneHlaseniXML({
      issuedInvoices,
      submitterData,
      year,
      quarter
    })

    // Use snapshot matching
    expect(result).toMatchSnapshot()
  })

  // ... (keep the test for empty invoices, adjusting the invoice type) ...
  it('should throw an error if no relevant EUR invoices are found', () => {
    const submitterData: SubmitterData = {
      dic: 'CZ123',
      typ_ds: 'F',
      prijmeni: 'Test',
      jmeno: 'User',
      naz_obce: 'City',
      ulice: 'Street',
      psc: '12345',
      stat: 'CZ',
      email: 'e@ma.il'
    }
    const issuedInvoices = [
      {
        id: '1',
        invoice_number: '2025-001',
        client_vat_no: 'CZ123',
        issue_date: '2025-03-01',
        native_total: 1000,
        currency: 'CZK'
      },
      {
        id: '2',
        invoice_number: '2025-002',
        native_total: 50,
        currency: 'USD'
      }
    ]

    expect(() =>
      generateSouhrnneHlaseniXML({
        issuedInvoices,
        submitterData,
        year: 2025,
        quarter: 1
      })
    ).toThrow('No relevant EUR invoices found for Souhrnné hlášení generation.')
  })
})
