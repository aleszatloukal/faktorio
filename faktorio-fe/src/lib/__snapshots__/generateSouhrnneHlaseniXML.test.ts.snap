// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`generateSouhrnneHlaseniXML > should generate correct XML using correct field names 1`] = `
"<?xml version="1.0" encoding="UTF-8"?>
<Pisemnost nazevSW="EPO MF ČR" verzeSW="45.5.1">
<DPHSHV verzePis="02.01">
<VetaD k_uladis="DPH" dokument="SHV" ctvrt="1" rok="2025" shvies_forma="R" d_poddp="21.04.2025" />
<VetaP dic="8807204153" typ_ds="F" prijmeni="Špác" jmeno="Jiří" naz_obce="SOKOLNICE" ulice="Třešňová" c_pop="823"  psc="66452" stat="ČESKÁ REPUBLIKA" />
<VetaR por_c_stran="1" c_rad="1" k_stat="DE" c_vat="360131145" k_pln_eu="3" pln_pocet="1" pln_hodnota="76388" />
<VetaR por_c_stran="2" c_rad="2" k_stat="DE" c_vat="987654321" k_pln_eu="3" pln_pocet="1" pln_hodnota="25000" />
<VetaR por_c_stran="3" c_rad="3" k_stat="FR" c_vat="12345678901" k_pln_eu="3" pln_pocet="1" pln_hodnota="50000" />
</DPHSHV>
</Pisemnost>"
`;
